File: ts_sdk/src/entities/git-repo.ts
import { APIEntity, registerEntity } from '../APIEntity';
import { IEntity } from '../IEntity';

export type GitProviderId = 'github' | 'gitlab' | 'bitbucket';

export interface IGitRepo extends IEntity {
  /** "github" — v1 only supports GitHub. */
  provider?: GitProviderId;
  owner?: string;
  /** Repo name (without owner). */
  name?: string;
  /** "owner/repo" — canonical display id. */
  full_name?: string;
  /** Repo's mainline branch, used for fallback labels. */
  default_branch?: string;
  /** The branch this shared location represents (the "current git location"). */
  branch?: string;
  /** Commit SHA at share time (snapshot, optional). */
  head_commit?: string | null;
  /** Public web URL (github.com/owner/repo). */
  html_url?: string;
  description?: string | null;
  private?: boolean;
  fork?: boolean;
}

/**
 * Shareable "git location" entity — represents a repo + branch the sender
 * wants the recipient to work on. Same wire pattern as Markdown/Spec:
 * created locally, attached to a FlowMessage as a `type_id` reference,
 * rendered as a chip on the recipient side, opens an accept-modal on click.
 *
 * Non-secret metadata only. For the workdir-bound git ops helper, see
 * `./git-workdir.ts` (`GitWorkdir`).
 */
@registerEntity
export class GitRepo extends APIEntity<GitRepo> implements IGitRepo {
  provider?: GitProviderId;
  owner?: string;
  name?: string;
  full_name?: string;
  default_branch?: string;
  branch?: string;
  head_commit?: string | null;
  html_url?: string;
  description?: string | null;
  private?: boolean;
  fork?: boolean;
  static type: string = 'git_repo';

  constructor(entity: Partial<IGitRepo> = {}) {
    super(entity);
    this.provider = entity.provider ?? 'github';
    this.owner = entity.owner ?? '';
    this.name = entity.name ?? '';
    this.full_name = entity.full_name ?? '';
    this.default_branch = entity.default_branch ?? 'main';
    this.branch = entity.branch ?? '';
    this.head_commit = entity.head_commit ?? null;
    this.html_url = entity.html_url ?? '';
    this.description = entity.description ?? null;
    this.private = entity.private ?? false;
    this.fork = entity.fork ?? false;
  }

  /** "owner/repo · branch" — what the chip label and modal header show. */
  get displayLabel(): string {
    if (this.full_name && this.branch) return `${this.full_name} · ${this.branch}`;
    if (this.full_name) return this.full_name;
    return this.name ?? '';
  }
}

File: flow_sdk/builtin/git_repo.py
"""GitRepo entity — a shareable pointer to a "git location" (repo + branch).

Sender attaches one of these to a conversation message; recipient sees a
chip in their FlowMessage bubble and opens an accept-modal that drives the
local clone / checkout / pull against the recipient's chosen project.

Fields are non-secret metadata mirroring the canonical ``RepoSummary`` shape
produced by ``flow_sdk/app/actions/repo_actions.py:list_user_repos`` plus a
``branch`` and ``head_commit`` snapshot taken at share time. No tokens, no
SSH keys, no local paths.

Identity: ``id`` is a fresh ``uuid4`` per materialization (Entity default
factory). Two senders sharing the same upstream repo produce two distinct
``GitRepo`` TypeIds — no deterministic discovery possible from a public
URL.
"""
from typing import ClassVar, Optional

from flow_sdk.api.api_types.api_field import APIField
from flow_sdk.core import Entity


class GitRepo(Entity):
    type: str = APIField(default="git_repo")

    # Provider + identity.
    provider: str = APIField(default="github")
    owner: str = APIField(default="")
    name: str = APIField(default="")
    full_name: str = APIField(default="")

    # The shared "location" — branch is the point-in-time picked branch;
    # default_branch is the repo's mainline (for fallback labels in the UI).
    default_branch: str = APIField(default="main")
    branch: str = APIField(default="")
    head_commit: Optional[str] = APIField(default=None)

    # Non-secret metadata; safe to share even for private repos (recipient
    # cannot clone without their own access).
    html_url: str = APIField(default="")
    description: Optional[str] = APIField(default=None)
    private: bool = APIField(default=False)
    fork: bool = APIField(default=False)

    _api_visible: ClassVar[bool] = True
    _icon: ClassVar[str] = "GitBranch"

    @property
    def display_name(self) -> str:
        if self.full_name and self.branch:
            return f"{self.full_name} · {self.branch}"
        if self.full_name:
            return self.full_name
        return self.name or ""
